File: build_all_data/terrain/code/build_slope.R
if (!require(terra)) install.packages('terra')
if (!require(sf)) install.packages('sf')
if (!require(tidyverse)) install.packages('tidyverse')
if (!require(R.utils)) install.packages('R.utils')
library(sf)
library(tidyverse)
library(terra)
library(R.utils)

rm(list=ls())

path <- "/home/dashaa/wildfire_politics/data/build_terrain"
input_path <- paste0(path,"/input/raw/copernicus")
output_path <- paste0(path,"/output")
#unzip gz files
# files <- list.files(input_path,pattern=".gz",full.names = T,recursive=T)
# for (file in files) {
#   gunzip(file)
# }
# files <- list.files(input_path,pattern=".tar",full.names = T,recursive=T)
# for (file in files) {
#   setwd(dirname(file))
#   untar(file)
#   unlink(file)
# }
slope_files <- list.files(input_path,pattern="slope.tif",full.names=T,recursive=T)
slope_rasts <- lapply(slope_files, rast)

incidents <- st_read("/home/dashaa/wildfire_politics/data/build_wildfire_reports/output/incidents_spatial.gpkg")
incidents <- subset(incidents,select=c(incident_id,final_acres))
incidents <- st_transform(incidents, "+proj=utm +zone=12 +datum=WGS84")
incidents$final_sq_m <- incidents$final_acres*4046.8564224
incidents$radius <- sqrt(incidents$final_sq_m/pi)
incidents <- st_buffer(incidents,incidents$radius)
incidents <- st_transform(incidents, st_crs(slope_rasts[[1]]))

incidents$mean_slope <- 0
incidents$max_slope <- NA
incidents$min_slope <- NA
incidents$total_cells <- 0

#now extract slope mean, max and min for each incident
for (slope_rast in slope_rasts) {
  
  slope_means <- terra::extract(slope_rast,incidents,mean,na.rm=T)
  slope_n_cells <- terra::extract(slope_rast,incidents,function(x) sum(!is.na(x)))
  
  added_means <- rowMeans(slope_n_cells[2]*slope_means[2])
  added_means[is.na(added_means)] <- 0
  
  incidents$total_cells <- incidents$total_cells + rowMeans(slope_n_cells[2])
  incidents$mean_slope <- incidents$mean_slope + added_means
  
  slope_maxs <- terra::extract(slope_rast,incidents,max,na.rm=T)
  incidents$max_slope <- pmax(incidents$max_slope,rowMeans(slope_maxs[2]),na.rm=T)
  
  slope_mins <- terra::extract(slope_rast,incidents,min,na.rm=T)
  incidents$min_slope <- pmin(incidents$min_slope,rowMeans(slope_mins[2]),na.rm=T)
}

incidents$mean_slope <- incidents$mean_slope/incidents$total_cells

incidents <- st_set_geometry(incidents,NULL)
incidents <- subset(incidents,select=c(incident_id,mean_slope,max_slope,min_slope,total_cells,final_acres))
write.csv(incidents,paste0(output_path,"/incidents_slope.csv"))

File: build_all_data/terrain/code/build_slope.py
#build slope features
import os
import geopandas as gpd
import rasterio
from rasterio import features
import numpy as np
import pandas as pd
from shapely.geometry import Point, Polygon

# Paths
path = "/home/dashaa/wildfire_politics/data/build_terrain"
input_path = os.path.join(path, "input/raw/copernicus")
output_path = os.path.join(path, "output")

# Read slope files
slope_files = [os.path.join(root, file)
               for root, _, files in os.walk(input_path)
               for file in files if file.endswith("slope.tif")]

# Read incidents data
incidents = gpd.read_file("/home/dashaa/wildfire_politics/data/build_wildfire_reports/output/incidents_spatial.gpkg")
incidents = incidents[['incident_id', 'final_acres']]
incidents = incidents.to_crs(epsg=32612)
incidents['final_sq_m'] = incidents['final_acres'] * 4046.8564224
incidents['radius'] = np.sqrt(incidents['final_sq_m'] / np.pi)
incidents['geometry'] = incidents.buffer(incidents['radius'])
incidents = incidents.to_crs(crs=rasterio.open(slope_files[0]).crs)

# Initialize slope columns
incidents['mean_slope'] = 0
incidents['max_slope'] = np.nan
incidents['min_slope'] = np.nan
incidents['total_cells'] = 0

# Extract slope data
for slope_file in slope_files:
    with rasterio.open(slope_file) as src:
        for idx, incident in incidents.iterrows():
            geometry = [incident['geometry']]
            masked, _ = rasterio.mask.mask(src, geometry, crop=True, nodata=np.nan)
            masked = masked[0]

            if masked.size > 0:
                mean_val = np.nanmean(masked)
                max_val = np.nanmax(masked)
                min_val = np.nanmin(masked)
                num_cells = np.sum(~np.isnan(masked))

                if not np.isnan(mean_val):
                    incidents.at[idx, 'mean_slope'] += mean_val * num_cells
                    incidents.at[idx, 'total_cells'] += num_cells
                if not np.isnan(max_val):
                    incidents.at[idx, 'max_slope'] = max(max_val, incidents.at[idx, 'max_slope'])
                if not np.isnan(min_val):
                    incidents.at[idx, 'min_slope'] = min(min_val, incidents.at[idx, 'min_slope'])

# Finalize mean slope calculation
incidents['mean_slope'] /= incidents['total_cells']

# Save to CSV
incidents = incidents[['incident_id', 'mean_slope', 'max_slope', 'min_slope', 'total_cells', 'final_acres']]
incidents.to_csv(os.path.join(output_path, "incidents_slope.csv"), index=False)

